A94-subchannel-otel-metrics.md

```diff
@@ -0,0 +1,128 @@
+grpc.subchannel.connection_attempts_failed (Old - grpc.lb.pick_first.connection_attempts_failed)       | Counter        | {attempt}       | grpc.target, grpc.lb.backend_service (optional), grpc.lb.locality (optional)                                   | Number of failed connection attempts.
+grpc.subchannel.open_connections                                                                       | UpDown Counter | {connection}    | grpc.target, grpc.lb.backend_service (optional), grpc.lb.locality (optional)                                   | Number of open connections.
+
+If we end up discarding connection attempts as we do with the “happy eyeballs”
+algorithm (as per
+[A61: IPv4 and IPv6 Dualstack Backend Support](A61-IPv4-IPv6-dualstack-backends.md)),
+we should not record the connection attempt or the disconnection.
+
+Implementations that have already implemented the pick-first metrics should give
+enough time for users to transition to the new metrics. For example,
+implementations should report both the old pick-first metrics and the new
+subchannel metrics for 2 releases, and then remove the old pick-first metrics.
+
+Label Name              | Disposition | Description
+----------------------- | ----------- | -----------
+grpc.target             | Required    | Indicates the target of the gRPC channel (defined in [A66](A66-otel-stats.md).)
+grpc.lb.backend_service | Optional    | The backend service to which the RPC was routed (defined in [A89](https://github.com/grpc/proposal/pull/471))
+grpc.lb.locality        | Optional    | The locality to which the traffic is being sent. This will be set to the resolver attribute passed down from the weighted_target policy, or the empty string if the resolver attribute is unset. (defined in [A78](A78-grpc-metrics-wrr-pf-xds.md))
+grpc.disconnect_error   | Optional    | Reason for disconnection
+
+List of allowed values for `grpc.disconnect_error` -
+
+Error string         | Description
+-------------------- | -----------
+GOAWAY <ERROR_CODE>  | HTTP2 GOAWAY frame with error code for example (“GOAWAY NO_ERROR”, “GOAWAY PROTOCOL_ERROR”, “GOAWAY ENHANCE_YOUR_CALM”). The list of error codes is available in [RFC 9113](https://www.rfc-editor.org/rfc/rfc9113.html#name-error-codes).
+subchannel shutdown  | The subchannel was shutdown. This can happen due to reasons such as the parent channel shutting down, channel becoming idle, the load balancing policy changing due to a resolver update or a change in list of endpoint addresses.
+connection reset     | Connection was reset (eg. ECONNRESET, WSAECONNERESET)
+connection timed out | Connection timed out, also includes connections closed due to gRPC keepalives.
+connection aborted   | Connection was aborted
```

Comment:
I don't remember discussing this one -- not sure if we didn't, or if I just forgot.

What exactly triggers this one?  How is it different from "connection reset"?

Reply:
I added "connection timed out", "connection reset" and "connection aborted" as the most common socket errors and implemented across platforms. There are some differences across platforms as to the exact implementation but overall - 
"connection reset" - a TCP RST was received
"connection timed out" - keepalive or tcp user timeout
"connection aborted" - The connection was aborted/closed by the host itself (software initiated - application/OS). This can happen due to reasons like resource exhaustion, unexpected/internal error.

Reply:
In C-core, how are we going to differentiate between these cases?  We would presumably need to expose the difference between them in the EventEngine API, and then we need to plumb that data through the transport and up to the subchannel.  How exactly will this work?

Reply:
EventEngine callbacks have a status parameter that we can use. The endpoint will annotate the status object with the correct socket error message. 

For the subchannel, in Core, we have connectivity state watchers on the transport that also have a status object that is part of the callback, so the same status object that can be propagated up to the subchannel that can decode the error message.

Reply:
discussed on vc